language/stdlib/modules/libra_system_2.move

```diff
@@ -0,0 +1,254 @@
+        // if(LibraTimestamp::is_genesis()) {
+        // return ()
+        // };
+        Transaction::assert(Transaction::sender() == 0xA550C18 || Transaction::sender() == 0x0, 1);
+        let validators = &mut borrow_global_mut<ValidatorSet>(0x1D8).validators;
+
+        let size = Vector::length(validators);
+        if (size == 0) {
+            return
+        };
+
+        let i = 0;
+        let configs_changed = false;
+        configs_changed = configs_changed || update_ith_validator_info_(validators, i);
+        loop {
+            i = i + 1;
+            if (i >= size) { break };
+            configs_changed = configs_changed || update_ith_validator_info_(validators, i);
+        };
+        Transaction::assert(configs_changed, 42);
+        emit_reconfiguration_event();
+    }
+
+    // Public getters
+
+    // Return true if addr is a current validator
+    public fun is_validator(addr: &address): bool acquires ValidatorSet { is_validator_(addr, &borrow_global<ValidatorSet>(0x1D8).validators) }
+
+    // Returns validator info, if the address is not a validator, aborts
+    public fun get_validator_info(addr: &address): ValidatorInfo acquires ValidatorSet {
+        let validator_set_ref = borrow_global<ValidatorSet>(0x1D8);
+        let i = 0;
+        Transaction::assert(get_validator_index_(&validator_set_ref.validators, addr, &mut i), 19);
+
+        let validator_info = Vector::borrow(&validator_set_ref.validators, i);
+        return *validator_info
+    }
+
+    // Return the size of the current validator set
+    public fun validator_set_size(): u64 acquires ValidatorSet {
+        Vector::length(&borrow_global<ValidatorSet>(0x1D8).validators)
+    }
+
+    // Private helper functions
+
+    // Emit a reconfiguration event. This function will be invoked by the genesis to generate the very first
+    // reconfiguration event.
+    fun emit_reconfiguration_event() acquires ValidatorSet {
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+
+        // Ensure only one reconfiguration event is emitted per block
+        let current_block_time = LibraTimestamp::now_microseconds();
+
+        // Abort the transaction if reconfiguration was already called within the same block
+        Transaction::assert(current_block_time > validator_set_ref.last_reconfiguration_time, 14);
+
+        validator_set_ref.last_reconfiguration_time = current_block_time;
+        LibraAccount::emit_event<ValidatorSetChangeEvent>(
+            &mut validator_set_ref.change_events,
+            ValidatorSetChangeEvent {
+                new_validator_set: *&validator_set_ref.validators,
+            });
+    }
+
+    // Get the index of the validator with address `addr` in `validators`.
+    // Aborts if `addr` is not a current validator
+    fun get_validator_index_(validators: &vector<ValidatorInfo>, addr: &address, resulting_index: &mut u64): bool {
```

Comment:
I thought the plan was to have an internal function that might return -1 if nothing found and >= 0 otherwise. then we can have is_validator_ return index >= 0; similarly, in get_validator_index_ it would abort < 0.
Prefer avoiding the mutable u64 :p

Reply:
Yes, that was my though originally, but there are no negative numbers (no i64 or i32) in move. That's why I had separate functions before. I can only see this way to combine them into one.

Reply:
oh gosh...... that's why you were asking about unsigned lol

Reply:
You also return a `vector<u64>` that's empty if the index is invalid--that might be a bit better?